taiga_contrib_ldap_auth/services.py

```diff
@@ -62,17 +67,29 @@ def ldap_login_func(request):
     # (or any other attribute)
     login_input = request.DATA.get('username', None)
     pass_input = request.DATA.get('password', None)
-
-    # TODO: sanitize before passing to LDAP server?
-    username, email, full_name = connector.login(login = login_input,
-                                                 password = pass_input)
-
-    user = register_or_update(username = username,
-                              email = email,
-                              full_name = full_name)
-
-    data = make_auth_response_data(user)
-    return data
+    try:
+        # TODO: make sure these fields are sanitized before passing to LDAP server!
+        username, email, full_name = connector.login(login = login_input, password = password_input)
+    except connector.LDAPUserLoginError as ldap_error:
```

Comment:
Ah, yes, I understand now. The purpose of "fallback" is to be tried on login mismatch, and not on network outage (which I thought was the case). So, it works as you expect and describe, - it's just not what I expected. In other words, that's my problem. :)

-----

I'll try to be verbose in explaining why this raised a flag for me. Below is a scenario different from "either LDAP user or local user, but not both". It stems from the fact that without additional changes to `taiga-front`, the "Change password" page is still available to users. So, a user can have a password in both LDAP and Taiga's local DB.

Suppose Alice has a LDAP user account, `alice`, with password `oldpass`. One day, she gets wind of it having been leaked, or phished out of her by Mallory, so she changes it (with some organisation-provided web tool) to `newpass` in LDAP.

However, Mallory has already logged into Taiga by now, and used the "Change password" page to set the "normal" password to `pwned`.

What happens now if Alice or Mallory try to log in with either `newpass` or `pwned`, during normal operation or a network outage (when Taiga can't reach the LDAP server)?

vvv net / pass -> | newpass                        | pwned
:---:             | :---:           	              | :---:
normal operation  | log-in via LDAP                | LDAPUserLoginError -> log-in via `"normal"`
outage            | `LDAPConnectionError` -> failure | `LDAPConnectionError` -> failure

In other words, logging in is now possible with _both_ passwords. If a fallback auth mechanism is enabled, the user must know to change the `"normal"` password, too - which is counter-intuitive. Alternatively, a mechanism must be implemented by the system administrators to update `"normal"` password on LDAP password change - which should be communicated to them somewhere, lest they accidentally give themselves a backdoor.

Basically, allowing a fallback introduces a synchronisation issue.

What I expected was the following - if no LDAP-to-`"normal"` sync is done:

vvv net / pass -> | newpass                                              | pwned
:---:             | :---:                                                | :---:
normal operation  | log-in via LDAP                                      | `LDAPUserLoginError` -> failure
outage            | `LDAPConnectionError` -> `LDAPUserLoginError` -> failure | `LDAPConnectionError` -> log-in via `"normal"`

The above, of course, is not much better. The backdoor is active during network outages only, but there's still a backdoor.

With some form of auto-sync:

vvv net / pass -> | newpass                                    | pwned
:---:             | :---:                                      | :---:
normal operation  | log-in via LDAP                            | `LDAPUserLoginError` -> failure
outage            | `LDAPConnectionError` -> log-in via `"normal"` | `LDAPConnectionError` -> `LDAPUserLoginError` -> failure

Other solutions are possible, like disabling the "Change password" page, or doing sanity checks, or something else entirely.

Maybe I'm missigng something here, too?..